File: rust/1-TwoSum/main.rs
use std::{collections::HashMap};

pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> 
{
    let mut indexes: HashMap<i32, i32> = HashMap::new();

    for(index, value) in nums.iter().enumerate()
    {
        if let Some(x) = indexes.get(&(target - *value))
        {
            return vec![index as i32, *x]
        }
        else
        {
            indexes.insert(*value, index as i32);
        }
    }

    return Vec::new();
}

fn main() 
{
    println!("{:?}", two_sum(vec![2,7,11,15], 9));
    println!("{:?}", two_sum(vec![3,2,4], 6));
    println!("{:?}", two_sum(vec![3,3], 6));
}

File: go/1-TwoSum/main.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	// We'll map a value to the index we saw it
	indexes := make(map[int]int)

	for i, value := range nums {
		// For any value, the other value we need is target-value
		if existing, found := indexes[target-value]; found {
			return []int{existing, i}
		} else {
			// Remember where we saw the value for later 
			indexes[value] = i
		}
	}

	return []int{}
}

func main() {
	fmt.Println(twoSum([]int{2, 7, 11, 15}, 9))
	fmt.Println(twoSum([]int{3, 2, 4}, 6))
	fmt.Println(twoSum([]int{3, 3}, 6))
}
